### DISCUSSION TITLE
How to use a custom dataset for Heterogeneous Graph Learning ?

### DISCUSSION THREAD

**Discussion by Justsecret123** [pyg-team/pytorch_geometric#8175] [Q&A]
Following the tutorial on Heteregeneous Graph Learning, I created my own dataset using Networks graphs  as follows :

```
dataset = []
for key, value in test.items():
    print(key, end="\r")
    # New graph 
    graph = from_networkx(value)
    #graph = HeteroData(graph)
    dataset.append(graph)

p_dataset = MyDataset(root="../Scripts/pytorch_dataset",data_list=dataset)
```

After successfully loading the created dataset using ```data=torch.load("../Scripts/pytorch_dataset/processed/dataset_pytorch_obj.pt")```,  I cannot seem to understand how can I get the metadata from the loaded dataset, at this step :  
<img width="996" alt="Screenshot 2023-10-11 at 01 13 04" src="https://github.com/pyg-team/pytorch_geometric/assets/50279622/bb33e6c3-952b-41f1-8d1f-6e09e80e6bf3">

```
model = GNN(hidden_channels=64, out_channels=4)
model = to_hetero(model, data.metadata, aggr='sum')
```

 It just returns an error : ```AttributeError: 'tuple' object has no attribute 'metadata'```.

How can I get the metadata from the generated .pt file ?  

Also, how can I define the x_dict since it also returns an error ? Should I modify the original Nx graphs to contain a feature named "x", and how ? 

Edit : Adding the "x" feature to the nodes on my original nx graphs produced no effect on the outcome, the x_dict remains empty. 

<img width="738" alt="Screenshot 2023-10-11 at 01 20 09" src="https://github.com/pyg-team/pytorch_geometric/assets/50279622/649bf209-1406-4c89-805a-226b2cf26710">

**Comment by rusty1s** [comment]
First of all, it looks like the data from disk is stored as a tuple (not sure why), so you may want to unpack it if you are only interested in the `Data` object, e.g., `torch.load(path)[0]`. Second, note that `from_networkx` does not support heterogeneous graphs yet (there is an open issue about this), and so the usage of heterogeneous GNNs is not applicable here. You can get around this by manually creating a `HeteroData` object if you definitely need heterogeneous GNNs.

**Reply by Justsecret123** [reply]
I managed to generate a homogeneous graph with the _from_networkx_ method. This is what the dataset looks like : 
-```data[0] = Data(x=[17], edge_index=[2, 14706], type=[774], value=[774], edge_type=[774], role=[774], num_nodes=13158)```
-```data[1].keys() = ['edge_index', 'type', 'value', 'edge_type', 'role']```

Neither data[0] nor data[1] have an attribute named _x_dict_, even though all nodes explicitly contain an feature _x_. Loading the first index (0) and trying to train the same model generates this error : ```AttributeError: 'GlobalStorage' object has no attribute 'x_dict'```.

**Reply by rusty1s** [reply]
`to_hetero` expects that you are working on a `HeteroData` object, which exposes `data.x_dict` and `data.edge_index_dict` functionality. It is not supported on homogeneous graphs.

**Reply by Justsecret123** [reply]
As stated earlier, I managed to create a homogeneous graph. I'm not using the `to_hetero` or `HeteroData` function anymore. However, the problem remains : there's no x_dict on data[0] nor data[1].

**Reply by rusty1s** [reply]
For homogeneous graphs, you can simply pass `data.x` and `data.edge_index` to your model. The concept of `x_dict` and `edge_index_dict` only exists on heterogeneous graphs, where you have node features and graph structure given for different node and edge types.

**Reply by Justsecret123** [reply]
This is the dataset loaded from disk :

<img width="1055" alt="Screenshot 2023-10-21 at 17 59 12" src="https://github.com/pyg-team/pytorch_geometric/assets/50279622/24128541-6378-4b46-aa22-a8e19fd657d3">

Where should I get the edge index from ? Moreover, there's no _x_ feature on data[0] nor data[1], which looks strange since all my graphs possess that feature. 

I think we would benefit from a complete example with a custom dataset (In Memory?) loaded from disk/from a list object.

**Reply by rusty1s** [reply]
Can you share some more details on how you processed the data? If the data does have node features before saving it, it should also have node features after loading it.

**Reply by Justsecret123** [reply]
My original Networkx.Digraph nodes contain three features : _x_, _type_, and _value_. Before saving the list of graphs as a (InMemory)dataset, I can retrieve the _x_ feature of a specific graph by accessing the list at a specific index. 

<img width="850" alt="Screenshot 2023-10-24 at 16 27 47" src="https://github.com/pyg-team/pytorch_geometric/assets/50279622/72f98d06-8019-434b-83be-c3fb1236b15c">

Unfortunately, when loading the dataset from disk, that particular _x_ feature just seems to be left out of the equation. The definition of the _MyDataset_ class is as follows :

```
class MyDataset(InMemoryDataset):
    def __init__(self, root, data_list, transform=None):
        self.data_list = data_list
        super().__init__(root, transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def processed_file_names(self):
        return "dataset_pytorch_obj.pt"

    def process(self):
        torch.save(self.collate(self.data_list), self.processed_paths[0])

```

**Reply by rusty1s** [reply]
Can you share a minimal example to reproduce? This would help me to track down the issue as otherwise I am out of ideas.

**Reply by Justsecret123** [reply]
As I cannot share data related to the project at the moment, it would be better if I provide some cues about how I generate the graphs instead. 

Starting from a dictionary object, I create the nodes and the edges as follows (ex: Emotion nodes) : 
```
# Generating the Emotion node
scene_graph.add_node(emotion, x=emotion, type="Emotion", value="")
# Generating the edge from Character to Emotion
scene_graph.add_edge(character, emotion, type="expresses", role="")
``` 

> **The here obtained graphs and their structure have already been verified and used with other libraries, the problem doesn't seem to come from here.**

 Then, the list of graphs is fed by looping over the generated graphs, then applying the _from_networkx_ function to each graph and finally appending it to the list. 

After this process, I am able to retrieve the _x_ feature on each graph (ex: data[1].x). Unfortunately, after casting that list into a PyTorch dataset, that feature just disappears. 

I could have used the list object as it if there was a tutorial or guide using a list of graphs as the starting point of a node/graph classification problem.
